Add IBC solomachine  client content (#1369)
cosmos/sdk-tutorials: academy/3-ibc/solomachine.md

* When updating the `ConsensusState` with a new public key or diversifier through submitting a `Header`.
* When submitting evidence of `Misbehaviour` to freeze a malicious client.
* When verifying packet commitments.

#### Signing data

The solomachine will sign over some data, `SignBytes`:

``` typescript
interface SignBytes {
  sequence: uint64
  timestamp: uint64  
  diversifier: string
  path: []byte
  data: []byte
}
```

and store a `Signature` including the data bytes and a timestamp:

``` typescript
interface Signature {
  data: []byte
  timestamp: uint64
}
```

This signature is then added to a message, e.g. the `Header`:

```typescript
interface Header {
  sequence: uint64 // deprecated
  timestamp: uint64
  signature: Signature
  newPublicKey: PublicKey
  newDiversifier: string
}
```

#### Client message signature verification

Remember from the section on light client development that the client can be updated (with new state or evidence of misbehavior) by submitting a `ClientMessage`. 

> The `ClientMessage` will be passed onto the client through a `MsgUpdateClient` submitted (generally by a relayer). The `02-client`'s [`UpdateClient`](https://github.com/cosmos/ibc-go/blob/v7.0.0/modules/core/02-client/keeper/client.go#L48) method will then handle the client message by using [these 4 methods on the `ClientState` interface](https://github.com/cosmos/ibc-go/blob/02-client-refactor-beta1/modules/core/exported/client.go#L98-L109):

Reviewer: ```suggestion
The `ClientMessage` will be passed onto the client by submitting a `MsgUpdateClient`. Although IBC messages are typically submitted by relayers, it can be advantageous to integrate a relayer directly into the solomachine's server for certain cases, such as with crypto.com's implementation. The `02-client`'s [`UpdateClient`](https://github.com/cosmos/ibc-go/blob/v7.0.0/modules/core/02-client/keeper/client.go#L48) method will then handle the client message by using [these 4 methods on the `ClientState` interface](https://github.com/cosmos/ibc-go/blob/02-client-refactor-beta1/modules/core/exported/client.go#L98-L109):
```

* `VerifyClientMessage`
* `CheckForMisbehaviour`
* `UpdateStateOnMisbehaviour`
* `UpdateState`

You can inspect the [solomachine specification](https://github.com/cosmos/ibc/tree/main/spec/client/ics-006-solo-machine-client) for more details, but take a look at the [`verifyClientMessage`](https://github.com/cosmos/ibc/tree/main/spec/client/ics-006-solo-machine-client#validity-predicate) where a `switch` statement is used to identify the cases of submitting a header or evidence of misbehavior.

You will notice in both cases some pseudo code that requires to perform the signature checks:

```typescript
assert(checkSignature(cs.consensusState.publicKey, signBytes, {msgType}.signature))
```

> Whereby the stored public key is used to check if the signature used to sign the `signBytes` was in fact the private key corresponding to it.

Reviewer 2: ```suggestion
The stored public key is used to check if the signature used to sign the `signBytes` was in fact the private key corresponding to it.
```

A solomachine is a standalone process that can interact with blockchains through IBC. It can store key information like signed messages and private keys but has no consensus algorithm of its own. The _solomachine client_ can be seen as a verification algorithm capable of authenticating messages sent from a solomachine. With solomachines, one can access the IBC transport layer and blockchains (including features built on them) within the Interchain without developing one's own blockchain. 

Anything from a web application hosted on a server, to a browser, to the mobile in your pocket is a solo machine. And these systems are capable of speaking IBC!

This is made possible using the IBC solo machine client.

<HighlightBox type=note>

Note that we have a solomachine on the one hand. This is the actual standalone machine (see the examples above) that is used to sign messages to interact with IBC-enabled chains.
<br/><br/>

> On the other hand, we have the _solomachine client_, which is the on-chain client allowing to verify the messages _sent by_ the solo machine on a remote counterparty chain.
> <br/><br/>
> The solomachine itself also still stores a light client representing the chain it wants to communicate with through IBC.
> <br/><br/>

Reviewer 3: ```suggestion
On the other hand, we have the _solomachine client_, which is the on-chain client allowing us to verify the messages _sent by_ the solo machine on a remote counterparty chain.
<br/><br/>
The solomachine itself still stores a light client representing the chain it wants to communicate with through IBC.
<br/><br/>
```

<br/><br/>
Because a solomachine is a lot less complex than a full fledged blockchain though, you will notice that a lot of the development work for the solomachine client is simplified.

</HighlightBox>

#### `ClientState`

The `ClientSate` is rather simple, it will contain the `ConsensusState` and a field to indicate whether the client is frozen:

``` typescript
interface ClientState {
  frozen: boolean
  consensusState: ConsensusState
}
```

It must of course also implement the `ClientState` methods defined in [the spec](https://github.com/cosmos/ibc/tree/main/spec/core/ics-002-client-semantics#clientstate)

#### `ConsensusState`

Next to the public key, the `ConsensusSate` also contains a timestamp and diversifier. The diversifier is an arbitrary string, chosen when the client is created, designed to allow the same public key to be re-used across different solo machine clients (potentially on different chains) without being considered misbehavior.

```typescript
interface ConsensusState {

>   sequence: uint64 // deprecated

Author: Thanks for checking @crodriguezvega I was a little confused by that when reading and hoped reviewers would offer clarity

>   sequence: uint64 // deprecated

Reviewer 4: @srdtrk Does this mean that `//deprecated` ought to be removed?

>   sequence: uint64 // deprecated

Reviewer 5: The `//deprecated` comment is fine, so we can leave it. But I opened a [PR](https://github.com/cosmos/ibc/pull/972) in the spec repo to update the ICS 06 spec to reflect that ibc-go has the `sequence` in the `ClientState` and not in the `ConsensusState`.

* When updating the `ConsensusState` with a new public key or diversifier through submitting a `Header`.
* When submitting evidence of `Misbehaviour` to freeze a malicious client.
* When verifying packet commitments.

#### Signing data

The solomachine will sign over some data, `SignBytes`:

``` typescript
interface SignBytes {
  sequence: uint64
  timestamp: uint64  
  diversifier: string
  path: []byte
  data: []byte
}
```

and store a `Signature` including the data bytes and a timestamp:

``` typescript
interface Signature {
  data: []byte
  timestamp: uint64
}
```

This signature is then added to a message, e.g. the `Header`:

```typescript
interface Header {
  sequence: uint64 // deprecated
  timestamp: uint64
  signature: Signature
  newPublicKey: PublicKey
  newDiversifier: string
}
```

#### Client message signature verification

Remember from the section on light client development that the client can be  updated (with new state or evidence of misbehavior) by submitting a `ClientMessage`. 

> The `ClientMessage` will be passed onto the client through a `MsgUpdateClient` submitted (generally by a relayer). The `02-client`'s [`UpdateClient`](https://github.com/cosmos/ibc-go/blob/v7.0.0/modules/core/02-client/keeper/client.go#L48) method will then handle the client message by using [these 4 methods on the `ClientState` interface](https://github.com/cosmos/ibc-go/blob/02-client-refactor-beta1/modules/core/exported/client.go#L98-L109):

Reviewer 4: ```suggestion
The `ClientMessage` will be passed onto the client by submitting a `MsgUpdateClient`. This message is generally submitted by a relayer, but not necessarily, as with crypto.com's for instance. The `02-client`'s [`UpdateClient`](https://github.com/cosmos/ibc-go/blob/v7.0.0/modules/core/02-client/keeper/client.go#L48) method will then handle the client message by using [these 4 methods on the `ClientState` interface](https://github.com/cosmos/ibc-go/blob/02-client-refactor-beta1/modules/core/exported/client.go#L98-L109):
```

  assert(checkSignature(cs.consensusState.publicKey, signBytes, header.signature))
}

function verifyMisbehaviour(misbehaviour: Misbehaviour) {
  clientState = provableStore.get("clients/{clientMsg.identifier}/clientState")
  s1 = misbehaviour.signatureOne
  s2 = misbehaviour.signatureTwo
  pubkey = clientState.consensusState.publicKey
  diversifier = clientState.consensusState.diversifier
  timestamp = clientState.consensusState.timestamp
  // assert that the signatures validate and that they are different
  sigBytes1 = SignBytes(
    sequence: misbehaviour.sequence,
    timestamp: s1.timestamp,
    diversifier: diversifier,
    path: s1.path,
    data: s1.data
  )
  sigBytes2 = SignBytes(
    sequence: misbehaviour.sequence,
    timestamp: s2.timestamp,
    diversifier: diversifier,
    path: s2.path,
    data: s2.data
  )
  // either the path or data must be different in order for the misbehaviour to be valid
  assert(s1.path != s2.path || s1.data != s2.data)
  assert(checkSignature(pubkey, sigBytes1, misbehaviour.signatureOne.signature))
  assert(checkSignature(pubkey, sigBytes2, misbehaviour.signatureTwo.signature))
}
```

</ExpansionPanel>

#### State signature verification

> Similarly, to verify if the signature that signed a state transition is valid you will remember that a light client has to implement `VerifyMembership` and `VerifyNonMembership` methods.

Reviewer 3: ```suggestion
Similarly, to verify if the signature that signed a state transition is valid, remember that a light client has to implement `VerifyMembership` and `VerifyNonMembership` methods.
```
